firedrake/slate/preconditioners/static_condensation.py

```diff
@@ -0,0 +1,240 @@
+
+        A KSP is created for the reduced system after
+        static condensation is applied.
+        """
+        from firedrake import (FunctionSpace, Function,
+                               TrialFunction, TestFunction,
+                               DirichletBC, interpolate)
+        from firedrake.assemble import (allocate_matrix,
+                                        create_assembly_callable)
+        from ufl.algorithms.replace import replace
+
+        # Extract python context
+        prefix = pc.getOptionsPrefix() + "static_condensation_"
+        _, P = pc.getOperators()
+        self.cxt = P.getPythonContext()
+
+        if not isinstance(self.cxt, ImplicitMatrixContext):
+            raise ValueError("Context must be an ImplicitMatrixContext")
+
+        test, trial = self.cxt.a.arguments()
+        V = test.function_space()
+        mesh = V.mesh()
+
+        if len(V) > 1:
+            raise ValueError("Cannot use this PC for mixed problems.")
+
+        if V.ufl_element().sobolev_space().name != "H1":
+            raise ValueError("Expecting an H1-conforming element.")
+
+        if not V.ufl_element().cell().is_simplex():
+            raise NotImplementedError("Only simplex meshes are implemented.")
+
+        top_dim = V.finat_element._element.ref_el.get_dimension()
+        if not V.finat_element.entity_dofs()[top_dim][0]:
+            raise RuntimeError("There are no interior dofs to eliminate.")
+
+        # We decompose the space into an interior part and facet part
+        interior_element = V.ufl_element()["interior"]
+        facet_element = V.ufl_element()["facet"]
+        V_int = FunctionSpace(mesh, interior_element)
+        V_facet = FunctionSpace(mesh, facet_element)
+
+        # Get transfer kernel for moving data
+        self._transfer_kernel = get_transfer_kernels({'h1-space': V,
+                                                      'interior-space': V_int,
+                                                      'facet-space': V_facet})
+
+        # Set up functions for the H1 functions and the interior/trace parts
+        self.trace_solution = Function(V_facet)
+        self.interior_solution = Function(V_int)
+        self.h1_solution = Function(V)
+        self.h1_residual = Function(V)
+        self.interior_residual = Function(V_int)
+        self.trace_residual = Function(V_facet)
+
+        # Collect BCs for the facet problem
+        bcs = []
+        for bc in self.cxt.row_bcs:
+            if isinstance(bc.function_arg, Function):
+                g = interpolate(bc.function_arg, V_facet)
+            else:
+                g = bc.function_arg
+            bcs.append(DirichletBC(V_facet, g, bc.sub_domain))
+
+        if bcs:
+            msg = ("Currently strong bcs are not handled correctly. "
+                   "The solver may still converge with this PC if an "
+                   "appropriate iterative method is used. ")
+            log(WARNING, msg)
```

Comment:
I chose to log a warning over raising an exception because methods using this PC (with Dirichlet conditions) still converge (consistently in 4 GMRES iterations). We can have this "feature" exposed or just raise a `NotImplementedError` until we can apply strong boundary conditions in Slate. Thoughts?